Hw3

awong-pdx opened the pull request:
# Code Changes I Made and Why
I added code to the files in the hw3 folder to complete the requirements for Homework 3.
  
### Exercise 01
I tried to use Bootstrap's grid system to layout the character images flexibly, with an `lg` breakpoint to display four images per row.  
The hover behavior was added in `styles.css`.

### Exercise 02
I used the `reduce()` array method to count the number of members for each house, and also implemented `cleanName()` to attempt to validate some of the incorrect data. I probably could refactor that function to use regexes instead of basic boolean checks for the various cases.   

### Exercise 03
I implemented Exercise 3 using create-react-app and the following modules listed in the assignment document:

- react-router-dom
- axios
- react-bootstrap
- react-chartjs-2

The project folder, `hw3-react`, separates components by page (`home`, `houses`, `search`, `navigation`).  
The app makes one API call in `CharactersRequest.js`, then attempts to store and read from the copied data in session storage.

***

# What Feedback I'm Looking For
Following the homework guidelines, I would appreciate any feedback on:
 - **Accessibility**
 - **Responsiveness**
 - **Airbnb style guide rules**
 - **Unnecessary code**
 - **Proper and consistent formatting**
 
I would love any feedback, and would really appreciate comments on the react aspects, since I'm still especially new to that. One thing I was unsure of was if there are better/more accepted ways to access and store API data.  
Lastly, I know there is an issue where new searches don't clear the results. While I'll keep working on it, feedback would be more than welcome!
***

#### Thank you for your feedback!

awong-pdx committed:
Tentatively finished 01-characters

awong-pdx committed:
Starting 02-charts

awong-pdx committed:
Tentatively finished 02-charts

Needs clean up

awong-pdx committed:
Cleaned up 02-charts, started 03-react

Started a new react app with
create-react-app
and imported:
react-router-dom,
axios,
react-bootstrap,
chartjs

awong-pdx committed:
Added routing with navbar

Added routing with react-router-dom
and a navbar with links to home, search, houses

awong-pdx committed:
Added search bar to search page

Search doesn't function, but component is from react-bs
Helpful source: https://react-bootstrap.netlify.app/docs/forms/form-text

awong-pdx committed:
Added search bar name input functionality

Used props and callback props to allow for
character name input in SearchBar

awong-pdx committed:
Adding API call to get GoT characters

CharactersRequest component calls API w/ axios and
stores array in sessionStorage

Ext. sources:
(https://stackoverflow.com/questions/63193114/how-do-i-call-a-function-
automatically-when-page-loads-up-in-react-js-in-2020)
(https://rapidapi.com/guides/persist-api-response)

awong-pdx committed:
Implemented rough character search w/ results

Added case-insensitive search for a full-name
character match.
Results don't properly clear when
a new search begins

awong-pdx committed:
Updated file structure

Moved components to new folders,
grouping by page/function

awong-pdx committed:
Added chart to Houses page

Reused most of the logic from 02-charts

awong-pdx committed:
Cleanup and style changes

awong-pdx committed:
More small styling updates

awong-pdx committed:
Removed custom chart bg colors

awong-pdx committed:
Create tasks.json

awong-pdx committed:
Edit gitignore

Add packages

awong-pdx committed:
Commit packages

awong-pdx committed:
Remove codesandbox file

HaliTran commented on hw3/01-characters.js:
```diff
@@ -1,4 +1,88 @@
+  newFigure.id = `img-${imgIndex}`;
+
+  newImg.className = 'character-img';
+  newImg.style.width = '250px';
+  newImg.style.height = '300px';
```
I would recommend setting the width and height attributes in the style.css file. To clean up the code a bit and separation of concerns since you are using bootstrap as well!

awong-pdx replied:
That's a good point I'll move that to css, thanks!

HaliTran commented on hw3/01-characters.js:
```diff
@@ -1,4 +1,88 @@
+};
+
+const createImg = function createGenericImgForDOM(rowIndex, imgIndex) {
+  const rowContainer = document.getElementById(`row-${rowIndex}`);
+  const newCol = document.createElement('div');
```
Can you elaborate on the rowContainer and newCol variables? I'm may be misinterpreting their purpose.

HaliTran replied:
After inspecting the code in the browser, it looks like you're using Bootstrap grid but on groups of pictures. Although this works just as intended, I would recommend simply using flexbox where you have the parent 'container' and each photo would be an item in that container. Then in the CSS file, you can target the parent container and display flex. Therefore, you can get the same results but with less JS code!

awong-pdx replied:
I think my reasoning behind that was that I wanted a clear cutoff between screen sizes, so that rows would be limited to 2 images on mobile/smaller screens, and 4 otherwise. But I think that's a good thing to bring up, it could definitely be cleaner.

HaliTran replied:
Ah I see. Yeah, I can see where you're coming from and why you decided to do that. I misinterpreted your intention and was assuming you simply wanted to display the images and I suggested flexbox to do that instead. Thanks for clarifying!

HaliTran commented on hw3/02-charts.js:
```diff
@@ -29,17 +29,68 @@ const borderColors = [
+  if (
+    cleanedName === ''
+    || cleanedName === 'None'
+    || cleanedName === 'Unkown'
```
Typo, I assume you meant 'Unknown'

HaliTran commented on hw3/hw3-react/src/search/SearchPage.js:
```diff
@@ -0,0 +1,58 @@
+    // this isn't working
+    setResults([]);
+    alert(`You searched for ${name}!`);
+    const storedCharacters = sessionStorage.getItem('characters');
```
Repeat: Although sessionStorage does retrieve the data information, I would consider passing the data that you receive from the API call to the SearchPage component so that this wouldn't be required. It aligns more with how React should function in the sense that you pass data to other components to use.

awong-pdx replied:
That sounds like a much more valid solution, thanks! I'll have to look into refactoring this portion of the code.

HaliTran commented on hw3/hw3-react/src/houses/HouseChart.js:
```diff
@@ -0,0 +1,92 @@
+  };
+
+  const createChart = function createChartDataFromSessionStorage() {
+    const storedCharacters = sessionStorage.getItem('characters');
```
Although sessionStorage does retrieve the data information, I would consider passing the data that you receive from the API call to the SearchPage component so that this wouldn't be required. It aligns more with how React should function in the sense that you pass data to other components to use.

awong-pdx replied:
Thanks, all the props feedback was really helpful! I refactored the way data was managed to follow your logic.

HaliTran commented on hw3/hw3-react/src/CharactersRequest.js:
```diff
@@ -0,0 +1,21 @@
+  useEffect(() => {
+    getCharacters()
+      .then((data) => {
+        sessionStorage.setItem('characters', JSON.stringify(data));
```
Separating the logic for the API request to retrieve the data make sense for modularity which is what I did as well. However, I learned from the TA that making the request in the 'parent' component  (which your case would be the AppRouter.js file) is more efficient because you can then pass the data as a prop when you call the other components like Search and HouseChart instead of using sessionStorage.

awong-pdx replied:
Yeah I learned a little more about the concepts about state management and best practices in react a bit too late lol. I'll get to fixing this following the logic of your feedback. Thanks!

HaliTran commented on hw3/hw3-react/src/navigation/AppRouter.js:
```diff
@@ -0,0 +1,20 @@
+import HousesPage from '../houses/HousesPage';
+
+function AppRouter() {
+  return (
```
Making the API call and using a useState hook in this function would be ideal because once you store the API response data (list of objects), you can pass it to the different components in your Route when it is rendered. For example, if the user hits the /search endpoint, rendering the SearchPage component would look like `<Route path="/houses" element={<SearchPage character={data} />}></Route>`. Then in your SearchPage.js file, since the data was passed in as argument, you can use that data to create the logic behind your search instead of sessionStorage.

awong-pdx replied:
Refactored

HaliTran commented on hw3/hw3-react/src/search/SearchPage.js:
```diff
@@ -0,0 +1,58 @@
+    characters.data.map((character) => {
+      if (name.toLowerCase() === character.fullName.toLowerCase()) {
+        alert(`There was a match for ${name}!`);
+        setResults([...results, character]);
```
Search function ALMOST works, the logic is basically there! Once you get a match on a character, storing that character's information in an object and using setResult should work. Then in your return statement, doing a conditional check on if the object exist, displaying that object's properties by returning in image tag essentially.

awong-pdx replied:
Your feedback was really helpful in getting me to think about what setResults() really does. In my refactor I used `array.filter()` instead and removed the `...results` in `setResults` to clear any previous matches.

HaliTran commented:
Hi Alessandra!

Great job on homework #3! I left some comments here and there on how I'd think you could improve the assignment. Most of them were geared towards the React portion and how you could refactor it to utilize props and structure where the API call was being made. I definitely struggle with React too but your implementation overall was very modular and readable! You were so close with the search function and I hope the comments I left could guide you down the right path! 

Happy coding!

huutribui reviewed (approved):
Hi Alessandra,
Hali pointed out what is needed. For the React portion, I think if you don't want to pass the data down from the parent, it would be a great idea to use useContext. It is very convenient and simple. 
Moreover, for the search page, it would be a great improvement if you allow multiple characters displayed if they meet the keyword (ex: search for "Stark" and all characters with Stark in their full name will show up). It can be done by using characters.data.filter() instead of map (SearchPage.js line 35) and render the new array with the same logic as the 01 exercise.

awong-pdx committed:
extract chart options, moved browser router call

awong-pdx committed:
finish moving browser router

awong-pdx committed:
moved character img sizing to css

awong-pdx committed:
refactor to pass props from parent app

awong-pdx committed:
fixed colors not rendering w/ dynamic data

awong-pdx committed:
refactor search characters

Fixed result update error,
added ability to
search for partial matches (last name)

awong-pdx commented:
> Hi Alessandra, Hali pointed out what is needed. For the React portion, I think if you don't want to pass the data down from the parent, it would be a great idea to use useContext. It is very convenient and simple. Moreover, for the search page, it would be a great improvement if you allow multiple characters displayed if they meet the keyword (ex: search for "Stark" and all characters with Stark in their full name will show up). It can be done by using characters.data.filter() instead of map (SearchPage.js line 35) and render the new array with the same logic as the 01 exercise.

Thanks for your feedback! I took it and implemented those changes in my refactor. I didn't use context this time, but I want to make sure I know how to use it in the future, so thanks for mentioning it!
